### ISSUE TITLE
There is not enough RAM during training

### ISSUE THREAD

**Issue by Pro100rus32** [ultralytics/ultralytics#8087]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train

### Bug

Despite disabling the `--cache` parameter, the learning process leads to a memory shortage error and stops the learning process in colab, and I get `^C`. This error persists when using more than 6,000 images in the free colab

### Environment

Ultralytics YOLOv8.1.10 🚀 Python-3.10.12 torch-2.1.0+cu121 CUDA:0 (Tesla T4, 15102MiB)
Setup complete ✅ (2 CPUs, 12.7 GB RAM, 26.3/78.2 GB disk)

### Minimal Reproducible Example

`yolo train model=yolov8s-cls.pt data=/content/datasets epochs=50 imgsz=640 batch=92` _#default cache=False_
**OR**
`yolo train model=yolov8s-cls.pt data=/content/datasets epochs=50 imgsz=640 batch=60 cache=False`

### Additional

```
Ultralytics YOLOv8.1.10 🚀 Python-3.10.12 torch-2.1.0+cu121 CUDA:0 (Tesla T4, 15102MiB)
engine/trainer: task=classify, mode=train, model=yolov8s-cls.pt, data=/content/datasets, epochs=50, time=None, patience=50, batch=70, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=train2, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0, auto_augment=randaugment, erasing=0.4, crop_fraction=1.0, cfg=None, tracker=botsort.yaml, save_dir=runs/classify/train2
train: /content/datasets/train... found 5875 images in 21 classes ✅ 
val: /content/datasets/val... found 2087 images in 21 classes ✅ 
test: /content/datasets/test... found 1 images in 1 classes: ERROR ❌️ requires 21 classes, not 1
2024-02-07 20:10:15.434284: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-02-07 20:10:15.434350: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-02-07 20:10:15.441880: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
Overriding model.yaml nc=1000 with nc=21

                   from  n    params  module                                       arguments                     
  0                  -1  1       928  ultralytics.nn.modules.conv.Conv             [3, 32, 3, 2]                 
  1                  -1  1     18560  ultralytics.nn.modules.conv.Conv             [32, 64, 3, 2]                
  2                  -1  1     29056  ultralytics.nn.modules.block.C2f             [64, 64, 1, True]             
  3                  -1  1     73984  ultralytics.nn.modules.conv.Conv             [64, 128, 3, 2]               
  4                  -1  2    197632  ultralytics.nn.modules.block.C2f             [128, 128, 2, True]           
  5                  -1  1    295424  ultralytics.nn.modules.conv.Conv             [128, 256, 3, 2]              
  6                  -1  2    788480  ultralytics.nn.modules.block.C2f             [256, 256, 2, True]           
  7                  -1  1   1180672  ultralytics.nn.modules.conv.Conv             [256, 512, 3, 2]              
  8                  -1  1   1838080  ultralytics.nn.modules.block.C2f             [512, 512, 1, True]           
  9                  -1  1    684821  ultralytics.nn.modules.head.Classify         [512, 21]                     
YOLOv8s-cls summary: 99 layers, 5107637 parameters, 5107637 gradients, 12.6 GFLOPs
Transferred 156/158 items from pretrained weights
TensorBoard: Start with 'tensorboard --logdir runs/classify/train2', view at http://localhost:6006/
AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...
AMP: checks passed ✅
train: Scanning /content/datasets/train... 5875 images, 0 corrupt: 100% 5875/5875 [00:00<?, ?it/s]
val: Scanning /content/datasets/val... 2087 images, 0 corrupt: 100% 2087/2087 [00:00<?, ?it/s]
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.000714, momentum=0.9) with parameter groups 26 weight(decay=0.0), 27 weight(decay=0.000546875), 27 bias(decay=0.0)
TensorBoard: model graph visualization added ✅
Image sizes 640 train, 640 val
Using 2 dataloader workers
Logging results to runs/classify/train2
Starting training for 50 epochs...

      Epoch    GPU_mem       loss  Instances       Size
       1/50      11.1G      3.134         70        640:   2% 2/84 [00:21<12:00,  8.79s/it]^C
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@Pro100rus32 it seems like you're encountering out-of-memory issues during training on Colab with a large dataset. Here are a few suggestions to mitigate this:

1. **Reduce Batch Size**: Lower the `batch` size to decrease memory usage.
2. **Image Size**: Decrease `imgsz` to reduce the amount of GPU memory required.
3. **Use AutoBatch**: Set `batch=-1` to automatically adjust the batch size based on your dataset and GPU capabilities.
4. **Gradual Loading**: If not already done, avoid using `cache=True` as it loads the entire dataset into RAM.

If the issue persists, consider using a platform with more RAM or splitting your dataset into smaller chunks. 🛠️

**Comment by NeuralAIM** [comment]
@glenn-jocher I also don't have enough 16GB of RAM with 6000 images and with batch=16. I do not understand why there is such a large memory consumption, and why it depends on the number of images in the dataset if the cache parameter=False/disk

**Comment by Burhan-Q** [comment]
@NeuralAIM the computer _has_ to open the images for each batch and (if enabled) execute augmentations. There will _always_ be RAM consumption during training and it will grow with additional epochs as more information is generated and stored in variables. This is the case with long running processes and python does automatic cleanup whenever it can. Most issues with respect to RAM utilization pertaining to Ultralytics will ultimately have a root cause based in PyTorch, since the `dataset` and `dataloader` classes for Ultralytics inherit from the PyTorch classes.

**Comment by Pro100rus32** [comment]
This only reduces the amount of GPU memory, but not RAM:
> 2. **Image Size**: Decrease `imgsz` to reduce the amount of GPU memory required.

How can I do this, @glenn-jocher?
> =splitting your dataset into smaller chunks. 🛠️

**Comment by glenn-jocher** [comment]
@Pro100rus32 it seems you're encountering out-of-memory issues during training on Colab with a large dataset. Here are a few suggestions to mitigate this:

1. **Reduce Batch Size**: Lower the `batch` size to reduce memory requirements per training step.
2. **Image Size**: Decrease `imgsz` to a smaller dimension to reduce the memory footprint.
3. **Gradient Accumulation**: If reducing the batch size impacts learning, consider using gradient accumulation to effectively increase the batch size without increasing memory usage.
4. **Use a Lighter Model**: If you're not already, consider using a lighter model like `yolov8n` instead of larger variants.
5. **Upgrade Colab**: Consider using Colab Pro for more memory if your dataset is large.

Try adjusting these parameters and monitor the memory usage to prevent the `^C` error, which indicates a manual interruption, possibly due to the system killing the process due to excessive memory usage.

**Comment by Pro100rus32** [comment]
> @Pro100rus32 it seems you're encountering out-of-memory issues during training on Colab with a large dataset. Here are a few suggestions to mitigate this:
> 
> 1. **Reduce Batch Size**: Lower the `batch` size to reduce memory requirements per training step.
> 2. **Image Size**: Decrease `imgsz` to a smaller dimension to reduce the memory footprint.
> 3. **Gradient Accumulation**: If reducing the batch size impacts learning, consider using gradient accumulation to effectively increase the batch size without increasing memory usage.
> 4. **Use a Lighter Model**: If you're not already, consider using a lighter model like `yolov8n` instead of larger variants.
> 5. **Upgrade Colab**: Consider using Colab Pro for more memory if your dataset is large.
> 
> Try adjusting these parameters and monitor the memory usage to prevent the `^C` error, which indicates a manual interruption, possibly due to the system killing the process due to excessive memory usage.
1. This only reduces the amount of GPU memory, but not RAM.
2. I did not notice a significant decrease in RAM consumption when changing imgsz from 640 to 224px
3. Isn't it used by default?
> Disable gradient accumulation: By default, YOLOv8 performs gradient accumulation, which can increase memory usage. You can disable this by using the `--no-accumulate` argument during training.
p.s. https://github.com/ultralytics/ultralytics/issues/3745#issuecomment-1636743244

4. Changing the model also does not significantly affect RAM consumption (I already use the s model, switching to the n model will not help much)
5. This will not help much, because it is planned to add a dataset of 25000+ of images, and in this case even colab pro will be powerless : )

**Comment by glenn-jocher** [comment]
@Pro100rus32 it seems like you're encountering out-of-memory issues during training on Colab with a large dataset. Here are a few suggestions that might help:

1. **Reduce Batch Size**: Lower the `batch` size to reduce memory consumption.
2. **Image Size**: Decrease `imgsz` to a smaller dimension if possible.
3. **Use CPU**: If you're still facing issues, try training on the CPU by setting `device=cpu`, although it will be slower.
4. **Gradual Loading**: Instead of loading the entire dataset into memory, load data gradually as needed during training.

If the issue persists, consider using Colab Pro for more memory or training on a local machine with sufficient RAM.

**Comment by NeuralAIM** [comment]
@glenn-jocher is GPT? 🤣

**Comment by glenn-jocher** [comment]
GlennPT 🤣

**Comment by glenn-jocher** [comment]
Sounds like you might need to invest in local or cloud hardware if you've tried the above steps and you're still out of RAM. Colab is really meant for playing around more than serious training.
